kotlin/workflow-ui-android/src/main/java/com/squareup/workflow/ui/backstack/BackStackContainer.kt

```diff
@@ -51,45 +60,45 @@ class BackStackContainer(
 
   private val showing: View? get() = if (childCount > 0) getChildAt(0) else null
 
-  fun takeScreens(
-    screens: Observable<out BackStackScreen<*>>,
-    viewRegistry: ViewRegistry
-  ) {
-    takeWhileAttached(screens.distinctUntilChanged { a, b -> a.key == b.key }) {
-      show(it, screens, viewRegistry)
-    }
-  }
+  private lateinit var registry: ViewRegistry
 
-  private fun show(
-    newScreen: BackStackScreen<*>,
-    screens: Observable<out BackStackScreen<*>>,
-    viewRegistry: ViewRegistry
-  ) {
-    showing?.let { if (it.backStackKey == newScreen.key) return }
+  private fun update(newValue: BackStackScreen<*>) {
+    // Existing view is of the right type, just update it.
+    showing
+        ?.takeIf { it.canShowValue(newValue.wrapped) }
+        ?.showValue(newValue.wrapped)
+        ?.also { return }
 
-    val updateTools = viewStateStack.prepareToUpdate(newScreen.key)
+    val updateTools = viewStateStack.prepareToUpdate(newValue.key)
+    val newView = registry.buildView(newValue.wrapped, this)
+        .apply { updateTools.setUpNewView(this) }
 
+    // Showing something already, transition with push or pop effect.
     showing
-        ?.let {
-          updateTools.saveOldView(it)
-          viewRegistry.getEffect(it.backStackKey, newScreen.key, updateTools.direction)
-              .execute(
-                  from = it,
-                  to = newScreen,
-                  screens = screens,
-                  viewRegistry = viewRegistry,
-                  container = this,
-                  setUpNewView = updateTools::setUpNewView,
-                  direction = updateTools.direction
-              )
+        ?.let { oldView ->
+          updateTools.saveOldView(oldView)
+
+          val newScene = Scene(this, newView)
+
+          val (outEdge, inEdge) = when (updateTools.direction) {
+            PUSH -> Gravity.START to Gravity.END
+            POP -> Gravity.END to Gravity.START
+          }
+
+          val outSet = TransitionSet()
+              .addTransition(Slide(outEdge).addTarget(oldView))
+              .addTransition(Fade(Fade.OUT))
+
+          val fullSet = TransitionSet()
+              .addTransition(outSet)
+              .addTransition(Slide(inEdge).excludeTarget(oldView, true))
+
+          TransitionManager.go(newScene, fullSet)
         }
-        ?: NoEffect.execute(
-            newScreen,
-            screens,
-            viewRegistry,
-            this,
-            updateTools::setUpNewView
-        )
+        ?.also { return }
```

Comment:
not important: could we use return@update here and above instead of the unqualified return?

Reply:
Good thought, will do.